From 0462b0dd11b26736f372e414fefc6a4e270868a1 Mon Sep 17 00:00:00 2001
From: Jun Zhang <jun.zhang@arm.com>
Date: Thu, 10 Aug 2023 16:18:26 +0800
Subject: [PATCH] Add mhu v3 driver for se_proxy SP

The driver for MHU V3 is based on the implementation in
trusted-firmware-m [0]. Add a build parameter to select
the version of MHU and the default version is MHU V2.
If we use MHU V3, then the parameter as below is needed.

-DPLAT_MHU_VERSION=3

0:
platform/ext/target/arm/rss/common/native_drivers/mhu_v3_x.c
platform/ext/target/arm/rss/common/native_drivers/mhu_v3_x.h
platform/ext/target/arm/rss/common/native_drivers/mhu_wrapper_v3_x.c

Repo: https://gitlab.arm.com/infra-solutions/reference-design/platsw/trusted-firmware-m/
Tag: RD-INFRA-2023.06.28

Upstream-Status: Pending
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
---
 .../drivers/arm/mhu_driver/component.cmake    |  18 +-
 platform/drivers/arm/mhu_driver/mhu.h         | 107 +++
 platform/drivers/arm/mhu_driver/mhu_v3_x.c    | 646 ++++++++++++++++++
 platform/drivers/arm/mhu_driver/mhu_v3_x.h    | 259 +++++++
 .../drivers/arm/mhu_driver/mhu_wrapper_v3_x.c | 308 +++++++++
 5 files changed, 1333 insertions(+), 5 deletions(-)
 create mode 100644 platform/drivers/arm/mhu_driver/mhu.h
 create mode 100644 platform/drivers/arm/mhu_driver/mhu_v3_x.c
 create mode 100644 platform/drivers/arm/mhu_driver/mhu_v3_x.h
 create mode 100644 platform/drivers/arm/mhu_driver/mhu_wrapper_v3_x.c

diff --git a/platform/drivers/arm/mhu_driver/component.cmake b/platform/drivers/arm/mhu_driver/component.cmake
index 77a5a50..544d75d 100644
--- a/platform/drivers/arm/mhu_driver/component.cmake
+++ b/platform/drivers/arm/mhu_driver/component.cmake
@@ -1,12 +1,20 @@
 #-------------------------------------------------------------------------------
-# Copyright (c) 2021, Arm Limited and Contributors. All rights reserved.
+# Copyright (c) 2023, Arm Limited and Contributors. All rights reserved.
 #
 # SPDX-License-Identifier: BSD-3-Clause
 #
 #-------------------------------------------------------------------------------
 
 # Add source files for using mhu driver
-target_sources(${TGT}
-	PRIVATE
-	"${CMAKE_CURRENT_LIST_DIR}/mhu_v2_x.c"
-)
+if(PLAT_MHU_VERSION EQUAL 3)
+	target_sources(${TGT}
+		PRIVATE
+		"${CMAKE_CURRENT_LIST_DIR}/mhu_v3_x.c"
+		"${CMAKE_CURRENT_LIST_DIR}/mhu_wrapper_v3_x.c"
+	)
+else()
+	target_sources(${TGT}
+		PRIVATE
+		"${CMAKE_CURRENT_LIST_DIR}/mhu_v2_x.c"
+	)
+endif()
diff --git a/platform/drivers/arm/mhu_driver/mhu.h b/platform/drivers/arm/mhu_driver/mhu.h
new file mode 100644
index 0000000..e9f407b
--- /dev/null
+++ b/platform/drivers/arm/mhu_driver/mhu.h
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022-2023 Arm Limited. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __MHU_H__
+#define __MHU_H__
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Generic MHU error enumeration types.
+ */
+enum mhu_error_t {
+    MHU_ERR_NONE                =  0,
+    MHU_ERR_NOT_INIT            = -1,
+    MHU_ERR_ALREADY_INIT        = -2,
+    MHU_ERR_UNSUPPORTED_VERSION = -3,
+    MHU_ERR_UNSUPPORTED         = -4,
+    MHU_ERR_INVALID_ARG         = -5,
+    MHU_ERR_BUFFER_TOO_SMALL    = -6,
+    MHU_ERR_GENERAL             = -7,
+};
+
+/**
+ * \brief Initializes sender MHU.
+ *
+ * \param[in] mhu_sender_dev        Pointer to the sender MHU.
+ *
+ * \return Returns mhu_error_t error code.
+ *
+ * \note This function must be called before mhu_send_data().
+ */
+enum mhu_error_t mhu_init_sender(void *mhu_sender_dev);
+
+/**
+ * \brief Initializes receiver MHU.
+ *
+ * \param[in] mhu_receiver_dev      Pointer to the receiver MHU.
+ *
+ * \return Returns mhu_error_t error code.
+ *
+ * \note This function must be called before mhu_receive_data().
+ */
+enum mhu_error_t mhu_init_receiver(void *mhu_receiver_dev);
+
+/**
+ * \brief Sends data over MHU.
+ *
+ * \param[in] mhu_sender_dev    Pointer to the sender MHU.
+ * \param[in] mhu_outband_base  Pointer to the shared memory.
+ * \param[in] send_buffer       Pointer to buffer containing the data to be
+ *                              transmitted.
+ * \param[in] size              Size of the data to be transmitted in bytes.
+ *
+ * \return Returns mhu_error_t error code.
+ *
+ * \note The send_buffer must be 4-byte aligned and its length must be at least
+ *       (4 - (size % 4)) bytes bigger than the data size to prevent buffer
+ *       over-reading.
+ */
+enum mhu_error_t mhu_send_data(void *mhu_sender_dev,
+                               void *mhu_outband_base,
+                               uint8_t *send_buffer,
+                               size_t size);
+
+/**
+ * \brief Receives data from MHU.
+ *
+ * \param[in]     mhu_receiver_dev  Pointer to the receiver MHU.
+ * \param[out]    receive_buffer    Pointer the buffer where to store the
+ *                                  received data.
+ * \param[in,out] size              As input the size of the receive_buffer,
+ *                                  as output the number of bytes received.
+ *                                  As a limitation, the size of the buffer
+ *                                  must be a multiple of 4.
+ *
+ * \return Returns mhu_error_t error code.
+ *
+ * \note The receive_buffer must be 4-byte aligned and its length must be a
+ *       multiple of 4.
+ */
+enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
+                                  uint8_t *receive_buffer,
+                                  size_t *size);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* __MHU_H__ */
diff --git a/platform/drivers/arm/mhu_driver/mhu_v3_x.c b/platform/drivers/arm/mhu_driver/mhu_v3_x.c
new file mode 100644
index 0000000..b640b3b
--- /dev/null
+++ b/platform/drivers/arm/mhu_driver/mhu_v3_x.c
@@ -0,0 +1,646 @@
+/*
+ * Copyright (c) 2023 Arm Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mhu_v3_x.h"
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "trace.h"
+
+/* Postbox control page */
+struct _mhu_v3_x_pbx_ctrl_reg_t {
+    /* Offset: 0x000 (R/ ) Postbox Block Identifier */
+    const volatile uint32_t mhu_blk_id;
+    /* Offset: 0x004 (R/ ) Reserved */
+    const volatile uint8_t reserved_0[0x10 - 0x04];
+    /* Offset: 0x010 (R/ ) Postbox Feature Support 0 */
+    const volatile uint32_t pbx_feat_spt0;
+    /* Offset: 0x014 (R/ ) Postbox Feature Support 1 */
+    const volatile uint32_t pbx_feat_spt1;
+    /* Offset: 0x018 (R/ ) Reserved */
+    const volatile uint8_t reserved_1[0x20 - 0x18];
+    /* Offset: 0x020 (R/ ) Postbox Doorbell Channel Configuration 0 */
+    const volatile uint32_t pbx_dbch_cfg0;
+    /* Offset: 0x024 (R/ ) Reserved */
+    const volatile uint8_t reserved_2[0x30 - 0x24];
+    /* Offset: 0x030 (R/ ) Postbox FIFO Channel Configuration 0 */
+    const volatile uint32_t pbx_ffch_cfg0;
+    /* Offset: 0x034 (R/ ) Reserved */
+    const volatile uint8_t reserved_3[0x40 - 0x34];
+    /* Offset: 0x040 (R/ ) Postbox Fast Channel Configuration 0 */
+    const volatile uint32_t pbx_fch_cfg0;
+    /* Offset: 0x044 (R/ ) Reserved */
+    const volatile uint8_t reserved_4[0x100 - 0x44];
+    /* Offset: 0x100 (R/W) Postbox control */
+    volatile uint32_t pbx_ctrl;
+    /* Offset: 0x164 (R/ ) Reserved */
+    const volatile uint8_t reserved_5[0x400 - 0x104];
+    /*
+     * Offset: 0x400 (R/ ) Postbox Doorbell Channel Interrupt Status n, where
+     * n is 0 - 3.
+     */
+    const volatile uint32_t pbx_dbch_int_st[4];
+    /*
+     * Offset: 0x410 (R/ ) Postbox FIFO Channel <n> Interrupt Status n, where n
+     * is 0 - 1.
+     */
+    const volatile uint32_t pbx_ffch_int_st[2];
+    /* Offset: 0x418 (R/ ) Reserved */
+    const uint8_t reserved_6[0xFC8 - 0x418];
+    /* Offset: 0xFC8 (R/ ) Postbox Implementer Identification Register */
+    const volatile uint32_t iidr;
+    /* Offset: 0xFCC (R/ ) Postbox Architecture Identification Register */
+    const volatile uint32_t aidr;
+    /*
+     * Offset: 0xFD0 (R/ ) Postbox Implementation Defined Identification
+     * Register n, where n is 0 - 11.
+     */
+    const volatile uint32_t impl_def_id[12];
+};
+
+/* Postbox Doorbell Channel Window Page <n>, where n is 0 - 127 */
+struct _mhu_v3_x_pbx_pdbcw_reg_t {
+    /* Offset: 0x000 (R/ ) Postbox Doorbell Channel Window Status */
+    const volatile uint32_t pdbcw_st;
+    /* Offset: 0x004 (R/ ) Reserved */
+    const uint8_t reserved_0[0xC - 0x4];
+    /* Offset: 0x00C ( /W) Postbox Doorbell Channel Window Set */
+    volatile uint32_t pdbcw_set;
+    /* Offset: 0x010 (R/ ) Postbox Doorbell Channel Window Interrupt Status */
+    const volatile uint32_t pdbcw_int_st;
+    /* Offset: 0x014 ( /W) Postbox Doorbell Channel Window Interrupt Clear */
+    volatile uint32_t pdbcw_int_clr;
+    /* Offset: 0x018 (R/W) Postbox Doorbell Channel Window Interrupt Enable */
+    volatile uint32_t pdbcw_int_en;
+    /* Offset: 0x01C (R/W) Postbox Doorbell Channel Window Control */
+    volatile uint32_t pdbcw_ctrl;
+};
+
+/* MHUv3 postbox structure */
+struct _mhu_v3_x_pbx {
+    /* Postbox Control */
+   struct _mhu_v3_x_pbx_ctrl_reg_t pbx_ctrl_page;
+   /* Postbox Doorbell Channel Window */
+   struct _mhu_v3_x_pbx_pdbcw_reg_t pdbcw_page;
+};
+
+/* Mailbox Control page */
+struct _mhu_v3_x_mbx_ctrl_reg_t {
+    /* Offset: 0x000 (R/ ) Mailbox Block Identifier */
+    const volatile uint32_t mhu_blk_id;
+    /* Offset: 0x004 (R/ ) Reserved */
+    const volatile uint8_t reserved_0[0x10 - 0x04];
+    /* Offset: 0x010 (R/ ) Mailbox Feature Support 0 */
+    const volatile uint32_t mbx_feat_spt0;
+    /* Offset: 0x014 (R/ ) Mailbox Feature Support 1 */
+    const volatile uint32_t mbx_feat_spt1;
+    /* Offset: 0x018 (R/ ) Reserved */
+    const volatile uint8_t reserved_1[0x20 - 0x18];
+    /* Offset: 0x020 (R/ ) Mailbox Doorbell Channel Configuration 0 */
+    const volatile uint32_t mbx_dbch_cfg0;
+    /* Offset: 0x024 (R/ ) Reserved */
+    const volatile uint8_t reserved_2[0x30 - 0x24];
+    /* Offset: 0x030 (R/ ) Mailbox FIFO Channel Configuration 0 */
+    const volatile uint32_t mbx_ffch_cfg0;
+    /* Offset: 0x034 (R/ ) Reserved */
+    const volatile uint8_t reserved_4[0x40 - 0x34];
+    /* Offset: 0x040 (R/ ) Mailbox Fast Channel Configuration 0 */
+    const volatile uint32_t mbx_fch_cfg0;
+    /* Offset: 0x044 (R/ ) Reserved */
+    const volatile uint8_t reserved_5[0x100 - 0x44];
+    /* Offset: 0x100 (R/W) Mailbox control */
+    volatile uint32_t mbx_ctrl;
+    /* Offset: 0x104 (R/ ) Reserved */
+    const volatile uint8_t reserved_6[0x140 - 0x104];
+    /* Offset: 0x140 (R/W) Mailbox Fast Channel control */
+    volatile uint32_t mbx_fch_ctrl;
+    /* Offset: 0x144 (R/W) Mailbox Fast Channel Group Interrupt Enable */
+    volatile uint32_t mbx_fcg_int_en;
+    /* Offset: 0x148 (R/ ) Reserved */
+    const volatile uint8_t reserved_7[0x400 - 0x148];
+    /*
+     * Offset: 0x400 (R/ ) Mailbox Doorbell Channel Interrupt Status n, where
+     * n = 0 - 3.
+     */
+    const volatile uint32_t mbx_dbch_int_st[4];
+    /*
+     * Offset: 0x410 (R/ ) Mailbox FIFO Channel Interrupt Status n, where
+     * n = 0 - 1.
+     */
+    const volatile uint32_t mbx_ffch_int_st[2];
+    /* Offset: 0x418 (R/ ) Reserved */
+    const volatile uint8_t reserved_8[0x470 - 0x418];
+    /* Offset: 0x470 (R/ ) Mailbox Fast Channel Group Interrupt Status */
+    const volatile uint32_t mbx_fcg_int_st;
+    /* Offset: 0x474 (R/ ) Reserved */
+    const volatile uint8_t reserved_9[0x480 - 0x474];
+    /*
+     * Offset: 0x480 (R/ ) Mailbox Fast Channel Group <n> Interrupt Status,
+     * where n = 0 - 31.
+     */
+    const volatile uint32_t mbx_fch_grp_int_st[32];
+    /* Offset: 0x500 (R/ ) Reserved */
+    const volatile uint8_t reserved_10[0xFC8 - 0x500];
+    /* Offset: 0xFC8 (R/ ) Mailbox Implementer Identification Register */
+    const volatile uint32_t iidr;
+    /* Offset: 0xFCC (R/ ) Mailbox Architecture Identification Register */
+    const volatile uint32_t aidr;
+    /*
+     * Offset: 0xFD0 (R/ ) Mailbox Implementation Defined Identification
+     * Register n, where n is 0 - 11.
+     */
+    const volatile uint32_t impl_def_id[12];
+};
+
+/* Mailbox Doorbell Channel Window Page <n>, where n is 0 - 127 */
+struct _mhu_v3_x_mbx_mdbcw_reg_t {
+    /* Offset: 0x000 (R/ ) Mailbox Doorbell Channel Window Status */
+    const volatile uint32_t mdbcw_st;
+    /* Offset: 0x004 (R/ ) Mailbox Doorbell Channel Window Status Masked */
+    const volatile uint32_t mdbcw_st_msk;
+    /* Offset: 0x008 ( /W) Mailbox Doorbell Channel Window Clear */
+    volatile uint32_t mdbcw_clr;
+    /* Offset: 0x00C (R/ ) Reserved */
+    const volatile uint8_t reserved_0[0x10 - 0x0C];
+    /* Offset: 0x010 (R/ ) Mailbox Doorbell Channel Window Mask Status */
+    const volatile uint32_t mdbcw_msk_st;
+    /* Offset: 0x014 ( /W) Mailbox Doorbell Channel Window Mask Set */
+    volatile uint32_t mdbcw_msk_set;
+    /* Offset: 0x018 ( /W) Mailbox Doorbell Channel Window Mask Clear */
+    volatile uint32_t mdbcw_msk_clr;
+    /* Offset: 0x01C (R/W) Mailbox Doorbell Channel Window Control */
+    volatile uint32_t mdbcw_ctrl;
+};
+
+/* MHUv3 mailbox structure */
+struct _mhu_v3_x_mbx {
+   /* Mailbox control */
+   struct _mhu_v3_x_mbx_ctrl_reg_t mbx_ctrl_page;
+   /* Mailbox Doorbell Channel Window */
+   struct _mhu_v3_x_mbx_mdbcw_reg_t mdbcw_page;
+};
+
+/* MHUv3 frame type */
+union _mhu_v3_x_frame_t {
+    /* Postbox Frame */
+    struct _mhu_v3_x_pbx pbx_frame;
+    /* Mailbox Frame */
+    struct _mhu_v3_x_mbx mbx_frame;
+};
+
+enum mhu_v3_x_error_t mhu_v3_x_driver_init(struct mhu_v3_x_dev_t *dev)
+{
+    uint32_t aidr = 0;
+    uint8_t mhu_major_rev;
+    union _mhu_v3_x_frame_t *p_mhu;
+
+    if (dev == NULL) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Return if already initialized */
+    if (dev->is_initialized) {
+        return MHU_V_3_X_ERR_NONE;
+    }
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    /* Read revision from MHU hardware */
+    if (dev->frame == MHU_V3_X_MBX_FRAME) {
+        aidr = p_mhu->mbx_frame.mbx_ctrl_page.aidr;
+    } else {
+        aidr = p_mhu->pbx_frame.pbx_ctrl_page.aidr;
+    }
+
+    /* Read the MHU Architecture Major Revision */
+    mhu_major_rev =
+        ((aidr & MHU_ARCH_MAJOR_REV_MASK) >> MHU_ARCH_MAJOR_REV_OFF);
+
+    /* Return error if the MHU major revision is not 3 */
+    if (mhu_major_rev != MHU_MAJOR_REV_V3) {
+        /* Unsupported MHU version */
+        return MHU_V_3_X_ERR_UNSUPPORTED_VERSION;
+    }
+
+    /* Read the MHU Architecture Minor Revision */
+    dev->subversion =
+        ((aidr & MHU_ARCH_MINOR_REV_MASK) >> MHU_ARCH_MINOR_REV_MASK);
+
+    /* Return error if the MHU minor revision is not 0 */
+    if (dev->subversion != MHU_MINOR_REV_3_0) {
+        /* Unsupported subversion */
+        return MHU_V_3_X_ERR_UNSUPPORTED_VERSION;
+    }
+
+    /* Initialize the Postbox/Mailbox to remain in operational state */
+    if (dev->frame == MHU_V3_X_MBX_FRAME) {
+        p_mhu->mbx_frame.mbx_ctrl_page.mbx_ctrl |= MHU_V3_OP_REQ;
+    } else{
+        p_mhu->pbx_frame.pbx_ctrl_page.pbx_ctrl |= MHU_V3_OP_REQ;
+    }
+
+    dev->is_initialized = true;
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_get_num_channel_implemented(
+     const struct mhu_v3_x_dev_t *dev,
+     enum mhu_v3_x_channel_type_t ch_type, uint8_t *num_ch)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+
+    if ((dev == NULL) || (num_ch == NULL)) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    /* Only doorbell channel is supported */
+    if (ch_type != MHU_V3_X_CHANNEL_TYPE_DBCH) {
+        return MHU_V_3_X_ERR_UNSUPPORTED;
+    }
+
+    /* Read the number of channels implemented in the MHU */
+    if(dev->frame == MHU_V3_X_PBX_FRAME) {
+        *num_ch = (p_mhu->pbx_frame.pbx_ctrl_page.pbx_dbch_cfg0 + 1);
+    } else {
+        *num_ch = (p_mhu->mbx_frame.mbx_ctrl_page.mbx_dbch_cfg0 + 1);
+    }
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_clear(const struct mhu_v3_x_dev_t *dev,
+     uint32_t channel, uint32_t mask)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+    struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
+    enum mhu_v3_x_error_t status;
+
+    if (dev == NULL) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized)) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    /* Only MBX can clear the Doorbell channel */
+    if(dev->frame != MHU_V3_X_MBX_FRAME) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+    mdbcw_reg = (struct _mhu_v3_x_mbx_mdbcw_reg_t *)
+        &(p_mhu->mbx_frame.mdbcw_page);
+
+    /* Clear the bits in the doorbell channel */
+    mdbcw_reg[channel].mdbcw_clr |= mask;
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_write(const struct mhu_v3_x_dev_t *dev,
+     uint32_t channel, uint32_t value)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+    struct _mhu_v3_x_pbx_pdbcw_reg_t *pdbcw_reg;
+    enum mhu_v3_x_error_t status;
+
+    if (dev == NULL) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized)) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    /* Only PBX can set the Doorbell channel value */
+    if (dev->frame != MHU_V3_X_PBX_FRAME) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    pdbcw_reg = (struct _mhu_v3_x_pbx_pdbcw_reg_t *)
+        &(p_mhu->pbx_frame.pdbcw_page);
+
+    /* Write the value to the doorbell channel */
+    pdbcw_reg[channel].pdbcw_set |= value;
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_read(const struct mhu_v3_x_dev_t *dev,
+     uint32_t channel, uint32_t *value)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+    enum mhu_v3_x_error_t status;
+
+    if ((dev == NULL) || (value == NULL)) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized)) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    if(dev->frame == MHU_V3_X_MBX_FRAME) {
+        struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
+        mdbcw_reg = (struct _mhu_v3_x_mbx_mdbcw_reg_t *)
+            &(p_mhu->mbx_frame.mdbcw_page);
+
+        /* Read the value from Mailbox Doorbell status register */
+        *value = mdbcw_reg[channel].mdbcw_st;
+    } else {
+        struct _mhu_v3_x_pbx_pdbcw_reg_t *pdbcw_reg;
+        pdbcw_reg = (struct _mhu_v3_x_pbx_pdbcw_reg_t *)
+            &(p_mhu->pbx_frame.pdbcw_page);
+
+        /* Read the value from Postbox Doorbell status register */
+        *value = pdbcw_reg[channel].pdbcw_st;
+    }
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_set(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel, uint32_t mask)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+    struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
+    enum mhu_v3_x_error_t status;
+
+    if (dev == NULL) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized) ) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    /* Doorbell channel mask in not applicable for PBX */
+    if(dev->frame != MHU_V3_X_MBX_FRAME) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    mdbcw_reg = (struct _mhu_v3_x_mbx_mdbcw_reg_t *)
+        &(p_mhu->mbx_frame.mdbcw_page);
+
+    /* Set the Doorbell channel mask */
+    mdbcw_reg[channel].mdbcw_msk_set = mask;
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_clear(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel, uint32_t mask)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+    struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
+    enum mhu_v3_x_error_t status;
+
+    if (dev == NULL) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized)) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    /* Doorbell channel mask in not applicable for PBX */
+    if(dev->frame != MHU_V3_X_MBX_FRAME) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    mdbcw_reg = (struct _mhu_v3_x_mbx_mdbcw_reg_t *)
+        &(p_mhu->mbx_frame.mdbcw_page);
+
+    /* Clear the Doorbell channel mask */
+    mdbcw_reg[channel].mdbcw_msk_clr = mask;
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_get(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel, uint32_t *mask_status)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+    struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
+    enum mhu_v3_x_error_t status;
+
+    if ((dev == NULL) || (mask_status == NULL)) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized)) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    /* Doorbell channel mask in not applicable for PBX */
+    if(dev->frame != MHU_V3_X_MBX_FRAME) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    mdbcw_reg = (struct _mhu_v3_x_mbx_mdbcw_reg_t *)
+        &(p_mhu->mbx_frame.mdbcw_page);
+
+    /* Save the Doorbell channel mask status */
+    *mask_status = mdbcw_reg[channel].mdbcw_msk_st;
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+void _mhu_v3_x_doorbell_interrupt_enable(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    if(dev->frame == MHU_V3_X_PBX_FRAME) {
+        struct _mhu_v3_x_pbx_pdbcw_reg_t *pdbcw_reg;
+        pdbcw_reg = (struct _mhu_v3_x_pbx_pdbcw_reg_t *)
+            &(p_mhu->pbx_frame.pdbcw_page);
+
+        /*
+         * Enable this doorbell channel to generate interrupts for transfer
+         * acknowledge events.
+         */
+        pdbcw_reg[channel].pdbcw_int_en = 0x1;
+
+        /*
+         * Enable this doorbell channel to contribute to the PBX combined
+         * interrupt.
+         */
+        pdbcw_reg[channel].pdbcw_ctrl = 0x1;
+    } else {
+        struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
+        mdbcw_reg = (struct _mhu_v3_x_mbx_mdbcw_reg_t *)
+            &(p_mhu->mbx_frame.mdbcw_page);
+
+        /*
+         * Enable this doorbell channel to contribute to the MBX combined
+         * interrupt.
+         */
+        mdbcw_reg[channel].mdbcw_ctrl = 0x1;
+    }
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_enable(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel,
+     enum mhu_v3_x_channel_type_t ch_type)
+{
+    enum mhu_v3_x_error_t status;
+
+    if (dev == NULL) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized)) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    /* Only doorbell channel is supported */
+    if (ch_type != MHU_V3_X_CHANNEL_TYPE_DBCH) {
+        return MHU_V_3_X_ERR_UNSUPPORTED;
+    }
+
+    /* Enable the doorbell channel interrupt */
+    _mhu_v3_x_doorbell_interrupt_enable(dev, channel);
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+void _mhu_v3_x_doorbell_interrupt_disable(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+
+    if(dev->frame == MHU_V3_X_PBX_FRAME) {
+        struct _mhu_v3_x_pbx_pdbcw_reg_t *pdbcw_reg;
+        pdbcw_reg = (struct _mhu_v3_x_pbx_pdbcw_reg_t *)
+            &(p_mhu->pbx_frame.pdbcw_page);
+
+        /* Clear channel transfer acknowledge event interrupt */
+        pdbcw_reg[channel].pdbcw_int_clr = 0x1;
+
+        /* Disable channel transfer acknowledge event interrupt */
+        pdbcw_reg[channel].pdbcw_int_en &= ~(0x1);
+
+        /*
+         * Disable this doorbell channel from contributing to the PBX combined
+         * interrupt.
+         */
+        pdbcw_reg[channel].pdbcw_ctrl &= ~(0x1);
+    } else {
+        struct _mhu_v3_x_mbx_mdbcw_reg_t *mdbcw_reg;
+        mdbcw_reg = (struct _mhu_v3_x_mbx_mdbcw_reg_t *)
+            &(p_mhu->mbx_frame.mdbcw_page);
+
+        /*
+         * Disable this doorbell channel from contributing to the MBX combined
+         * interrupt.
+         */
+        mdbcw_reg[channel].mdbcw_ctrl &= ~(0x1);
+    }
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_disable(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel,
+     enum mhu_v3_x_channel_type_t ch_type)
+{
+    enum mhu_v3_x_error_t status;
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized)) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    /* Only doorbell channel is supported */
+    if (ch_type != MHU_V3_X_CHANNEL_TYPE_DBCH) {
+        return MHU_V_3_X_ERR_UNSUPPORTED;
+    }
+
+    /* Disable the doorbell channel interrupt */
+    _mhu_v3_x_doorbell_interrupt_disable(dev, channel);
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+void _mhu_v3_x_doorbell_interrupt_clear(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel)
+{
+    union _mhu_v3_x_frame_t *p_mhu;
+    struct _mhu_v3_x_pbx_pdbcw_reg_t *pdbcw_reg;
+
+    p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+    pdbcw_reg = (struct _mhu_v3_x_pbx_pdbcw_reg_t *)&(p_mhu->
+            pbx_frame.pdbcw_page);
+
+    /* Clear channel transfer acknowledge event interrupt */
+    pdbcw_reg[channel].pdbcw_int_clr |= 0x1;
+}
+
+enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_clear(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel,
+     enum mhu_v3_x_channel_type_t ch_type)
+{
+    union _mhu_v3_x_frame_t *p_mhu = (union _mhu_v3_x_frame_t *)dev->base;
+    enum mhu_v3_x_error_t status;
+
+    /* Check if driver has been initialized */
+    if (!(dev->is_initialized)) {
+        return MHU_V_3_X_ERR_NOT_INIT;
+    }
+
+    /* Only doorbell channel is supported */
+    if (ch_type != MHU_V3_X_CHANNEL_TYPE_DBCH) {
+        return MHU_V_3_X_ERR_UNSUPPORTED;
+    }
+
+    /*
+     * Only postbox doorbell channel transfer acknowledge interrupt can be
+     * cleared manually.
+     */
+    if(dev->frame != MHU_V3_X_PBX_FRAME) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    /* Clear the postbox doorbell channel transfer acknowledge event */
+    _mhu_v3_x_doorbell_interrupt_clear(dev, channel);
+
+    return MHU_V_3_X_ERR_NONE;
+}
diff --git a/platform/drivers/arm/mhu_driver/mhu_v3_x.h b/platform/drivers/arm/mhu_driver/mhu_v3_x.h
new file mode 100644
index 0000000..d14946c
--- /dev/null
+++ b/platform/drivers/arm/mhu_driver/mhu_v3_x.h
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2023 Arm Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * \file mhu_v3_x.h
+ * \brief Driver for ARM Message Handling Unit version 3.0
+ */
+
+#ifndef __MHU_V3_X_H__
+#define __MHU_V3_X_H__
+
+#include <stdint.h>
+#include <stdbool.h>
+#include "mhu.h"
+//#include "host_base_address.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* MHU Architecture Major Revision 3 */
+#define MHU_MAJOR_REV_V3 (0x2u)
+/* MHU Architecture Minor Revision 0 */
+#define MHU_MINOR_REV_3_0 (0x0u)
+
+/* MHU Architecture Major Revision offset */
+#define MHU_ARCH_MAJOR_REV_OFF (0x4u)
+/* MHU Architecture Major Revision mask */
+#define MHU_ARCH_MAJOR_REV_MASK (0xfu << MHU_ARCH_MAJOR_REV_OFF)
+
+/* MHU Architecture Minor Revision offset */
+#define MHU_ARCH_MINOR_REV_OFF (0x0u)
+/* MHU Architecture Minor Revision mask */
+#define MHU_ARCH_MINOR_REV_MASK (0xfu << MHU_ARCH_MINOR_REV_OFF)
+
+/* MHUv3 PBX/MBX Operational Request offset */
+#define MHU_V3_OP_REQ_OFF (0u)
+/* MHUv3 PBX/MBX Operational Request */
+#define MHU_V3_OP_REQ (1u << MHU_V3_OP_REQ_OFF)
+
+#ifndef ALIGN_UP
+#define ALIGN_UP(num, align)    (((num) + ((align) - 1)) & ~((align) - 1))
+#endif
+
+/*
+ * MHUv3 Wrapper utility macros
+ */
+#define IS_ALIGNED(val, align) (val == ALIGN_UP(val, align))
+
+/* Flag Value used for AP<->RSS specific doorbell communication */
+#define MHU3_PBX_DBCH_FLAG_AP_COMMS     UINT32_C(0x2)
+
+/* Size of shared outband message buffer in bytes */
+#define MHU3_OUTBAND_BUF_SIZE UINT64_C(0x1000)
+#define MHU3_OUTBAND_BUF_HEADER_SIZE UINT64_C(0x4)
+
+/* Base of shared outband mesage buffer */
+//#define MHU3_OUTBAND_BUF_BASE HOST_AP_RSS_MAILBOX_BASE_S
+
+/**
+ * \brief MHUv3 error enumeration types.
+ */
+enum mhu_v3_x_error_t {
+    /* No error */
+    MHU_V_3_X_ERR_NONE =  0,
+    /* MHU driver not initialized */
+    MHU_V_3_X_ERR_NOT_INIT,
+    /* MHU Revision not supported error */
+    MHU_V_3_X_ERR_UNSUPPORTED_VERSION,
+    /* Operation not supported */
+    MHU_V_3_X_ERR_UNSUPPORTED,
+    /* Invalid parameter */
+    MHU_V_3_X_ERR_INVALID_PARAM,
+    /* General MHU driver error */
+    MHU_V_3_X_ERR_GENERAL,
+};
+
+/* Types of MHUv3 channels */
+enum mhu_v3_x_channel_type_t {
+    /* Doorbell channel */
+    MHU_V3_X_CHANNEL_TYPE_DBCH = (0x0u),
+    /* Channel type count */
+    MHU_V3_X_CHANNEL_TYPE_COUNT,
+};
+
+/* Types of the MHUv3 frames */
+enum mhu_v3_x_frame_t {
+    /* MHUv3 postbox frame */
+    MHU_V3_X_PBX_FRAME = (0x0u),
+    /* MHUv3 mailbox frame */
+    MHU_V3_X_MBX_FRAME = (0x1u),
+};
+
+/* MHUv3 device structure */
+struct mhu_v3_x_dev_t {
+    /* Base address of the MHUv3 frame */
+    //const uintptr_t base;
+    uintptr_t base;
+    /* Type of the MHUv3 frame */
+    enum mhu_v3_x_frame_t frame;
+    /* Minor revision of the MHUv3 */
+    uint32_t subversion;
+    /* Flag to indicate if the MHUv3 is initialized */
+    bool is_initialized;
+};
+
+size_t mhu_get_max_message_size(void);
+
+/**
+ * \brief Initializes the MHUv3.
+ *
+ * \param[in] dev    MHU device struct \ref mhu_v3_x_dev_t
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_driver_init(struct mhu_v3_x_dev_t *dev);
+
+/**
+ * \brief Returns the number of channels implemented.
+ *
+ * \param[in]  dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in]  ch_type    MHU channel type \ref mhu_v3_x_channel_type_t
+ * \param[out] num_ch     Pointer to the variable that will store the value.
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_get_num_channel_implemented(
+     const struct mhu_v3_x_dev_t *dev,
+     enum mhu_v3_x_channel_type_t ch_type, uint8_t *num_ch);
+
+/**
+ * \brief Clears the doorbell channel.
+ *
+ * \param[in] dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in] channel    Channel number.
+ * \param[in] mask       Mask to be used when clearing the channel.
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_clear(const struct mhu_v3_x_dev_t *dev,
+     uint32_t channel, uint32_t mask);
+
+/**
+ * \brief Write value to a doorbell channel.
+ *
+ * \param[in] dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in] channel    Doorbell channel number.
+ * \param[in] value      Value to be written to the channel.
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_write(const struct mhu_v3_x_dev_t *dev,
+     uint32_t channel, uint32_t value);
+
+/**
+ * \brief Read value from a doorbell channel.
+ *
+ * \param[in]  dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in]  channel    Doorbell channel number.
+ * \param[out] value      Pointer to the variable that will store the value.
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_read(const struct mhu_v3_x_dev_t *dev,
+     uint32_t channel, uint32_t *value);
+
+/**
+ * \brief Set bits in a doorbell channel mask.
+ *
+ * \param[in] dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in] channel    Doorbell channel number.
+ * \param[in] mask       Mask to be set over the doorbell channel.
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_set(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel, uint32_t mask);
+
+/**
+ * \brief Clear bits in a doorbell channel mask.
+ *
+ * \param[in] dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in] channel    Doorbell channel number.
+ * \param[in] mask       Mask to be cleared over the doorbell channel.
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_clear(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel, uint32_t mask);
+
+/**
+ * \brief Get the mask of a doorbell channel.
+ *
+ * \param[in]  dev            MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in]  channel        Doorbell channel number.
+ * \param[out] mask_status    Pointer to the variable that will store the value.
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_doorbell_mask_get(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel, uint32_t *mask_status);
+
+/**
+ * \brief Enable the channel interrupt.
+ *
+ * \param[in]  dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in]  channel    Channel number.
+ * \param[in]  ch_type    MHU channel type \ref mhu_v3_x_channel_type_t
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_enable(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel,
+     enum mhu_v3_x_channel_type_t ch_type);
+/**
+ * \brief Disable the channel interrupt.
+ *
+ * \param[in]  dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in]  channel    Channel number.
+ * \param[in]  ch_type    MHU channel type \ref mhu_v3_x_channel_type_t
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_disable(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel,
+     enum mhu_v3_x_channel_type_t ch_type);
+
+/**
+ * \brief Clear the channel interrupt.
+ *
+ * \param[in]  dev        MHU device struct \ref mhu_v3_x_dev_t
+ * \param[in]  channel    Channel number.
+ * \param[in]  ch_type    MHU channel type \ref mhu_v3_x_channel_type_t
+ *
+ * \return Returns error code as specified in \ref mhu_v3_x_error_t
+ */
+enum mhu_v3_x_error_t mhu_v3_x_channel_interrupt_clear(
+     const struct mhu_v3_x_dev_t *dev, uint32_t channel,
+     enum mhu_v3_x_channel_type_t ch_type);
+
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* __MHU_V3_X_H__ */
diff --git a/platform/drivers/arm/mhu_driver/mhu_wrapper_v3_x.c b/platform/drivers/arm/mhu_driver/mhu_wrapper_v3_x.c
new file mode 100644
index 0000000..74eddec
--- /dev/null
+++ b/platform/drivers/arm/mhu_driver/mhu_wrapper_v3_x.c
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2023 Arm Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mhu_v3_x.h"
+#include <string.h>
+
+static enum mhu_error_t error_mapping_to_mhu_error_t(enum mhu_v3_x_error_t err)
+{
+    switch (err) {
+    case MHU_V_3_X_ERR_NONE:
+        return MHU_ERR_NONE;
+    case MHU_V_3_X_ERR_NOT_INIT:
+        return MHU_ERR_NOT_INIT;
+    case MHU_V_3_X_ERR_UNSUPPORTED_VERSION:
+        return MHU_ERR_UNSUPPORTED_VERSION;
+    case MHU_V_3_X_ERR_UNSUPPORTED:
+        return MHU_ERR_UNSUPPORTED;
+    case MHU_V_3_X_ERR_INVALID_PARAM:
+        return MHU_ERR_INVALID_ARG;
+    default:
+        return MHU_ERR_GENERAL;
+    }
+}
+
+enum mhu_error_t mhu_init_sender(void *mhu_sender_dev)
+{
+    enum mhu_v3_x_error_t mhu_err;
+    struct mhu_v3_x_dev_t *dev = mhu_sender_dev;
+    uint8_t num_ch;
+    uint32_t ch;
+
+    if (dev == NULL) {
+        return error_mapping_to_mhu_error_t(MHU_V_3_X_ERR_INVALID_PARAM);
+    }
+    /* Initialize MHUv3 */
+    mhu_err = mhu_v3_x_driver_init(dev);
+    if (mhu_err != MHU_V_3_X_ERR_NONE) {
+        return error_mapping_to_mhu_error_t(mhu_err);
+    }
+
+    /* This wrapper requires at least 1 channel to be implemented */
+    mhu_err = mhu_v3_x_get_num_channel_implemented(dev,
+            MHU_V3_X_CHANNEL_TYPE_DBCH, &num_ch);
+    if (mhu_err != MHU_V_3_X_ERR_NONE) {
+        goto fail;
+    } else if (num_ch == 0) {
+        mhu_err = MHU_V_3_X_ERR_UNSUPPORTED;
+        goto fail;
+    }
+
+    /*
+     * The sender polls the postbox doorbell channel window status register to
+     * get notified about successful transfer. So, disable the doorbell
+     * channel's contribution to postbox combined interrupt.
+     *
+     * Also, clear and disable the postbox doorbell channel transfer acknowledge
+     * interrupt.
+     */
+    for (ch = 0; ch < num_ch; ++ch) {
+        mhu_err = mhu_v3_x_channel_interrupt_disable(dev, ch,
+                MHU_V3_X_CHANNEL_TYPE_DBCH);
+        if (mhu_err != MHU_V_3_X_ERR_NONE) {
+            goto fail;
+        }
+    }
+
+    return MHU_ERR_NONE;
+
+fail:
+    return error_mapping_to_mhu_error_t(mhu_err);
+}
+
+enum mhu_error_t mhu_init_receiver(void *mhu_receiver_dev)
+{
+    enum mhu_v3_x_error_t err;
+    struct mhu_v3_x_dev_t *dev = mhu_receiver_dev;
+    uint32_t ch;
+    uint8_t num_ch;
+
+    if (dev == NULL) {
+        return error_mapping_to_mhu_error_t(MHU_V_3_X_ERR_INVALID_PARAM);
+    }
+
+    /* Initialize MHUv3 */
+    err = mhu_v3_x_driver_init(dev);
+    if (err != MHU_V_3_X_ERR_NONE) {
+        return error_mapping_to_mhu_error_t(err);
+    }
+
+    /* This wrapper requires at least 1 channel to be implemented */
+    err = mhu_v3_x_get_num_channel_implemented(dev, MHU_V3_X_CHANNEL_TYPE_DBCH,
+            &num_ch);
+    if (err != MHU_V_3_X_ERR_NONE) {
+        goto fail;
+    } else if (num_ch == 0) {
+        err = MHU_V_3_X_ERR_UNSUPPORTED;
+        goto fail;
+    }
+
+    for (ch = 0; ch < num_ch; ++ch) {
+        /* Clear the doorbell channel */
+        err = mhu_v3_x_doorbell_clear(dev, ch, UINT32_MAX);
+        if (err != MHU_V_3_X_ERR_NONE) {
+            goto fail;
+        }
+
+        /* Unmask doorbell channel interrupt */
+        err = mhu_v3_x_doorbell_mask_clear(dev, ch, UINT32_MAX);
+        if (err != MHU_V_3_X_ERR_NONE) {
+            goto fail;
+        }
+
+        /*
+         * Enable the doorbell channel's contribution to mailbox combined
+         * interrupt.
+         */
+        err = mhu_v3_x_channel_interrupt_enable(dev, ch,
+                MHU_V3_X_CHANNEL_TYPE_DBCH);
+        if (err != MHU_V_3_X_ERR_NONE) {
+            goto fail;
+        }
+    }
+
+    return MHU_ERR_NONE;
+
+fail:
+    return error_mapping_to_mhu_error_t(err);
+}
+
+size_t mhu_get_max_message_size(void)
+{
+    /*
+     * First 4 bytes represent the MHU msg header. Rest of buffer
+     * holds the actual message.
+     */
+    return (MHU3_OUTBAND_BUF_SIZE - MHU3_OUTBAND_BUF_HEADER_SIZE);
+}
+
+static int validate_buffer_params(uintptr_t buf_addr, size_t buf_size)
+{
+    if ((buf_addr == 0) || (!IS_ALIGNED(buf_addr, 4)) ||
+            (!IS_ALIGNED(buf_size, 4)) ||
+            buf_size > mhu_get_max_message_size()) {
+        return MHU_V_3_X_ERR_INVALID_PARAM;
+    }
+
+    return MHU_V_3_X_ERR_NONE;
+}
+
+static enum mhu_v3_x_error_t signal_and_wait_for_clear(
+        struct mhu_v3_x_dev_t *dev, uint32_t ch, uint32_t value)
+{
+    enum mhu_v3_x_error_t err = MHU_V_3_X_ERR_NONE;
+    uint32_t read_val;
+
+    /* Wait for any pending acknowledgement from transmitter side */
+    do {
+        err = mhu_v3_x_doorbell_read(dev, ch, &read_val);
+        if (err != MHU_V_3_X_ERR_NONE) {
+            goto fail;
+        }
+    } while ((read_val & value) == value);
+
+    /* Trigger the doorbell */
+    err = mhu_v3_x_doorbell_write(dev, ch, value);
+    if (err != MHU_V_3_X_ERR_NONE) {
+        goto fail;
+    }
+
+    /* Wait until transmitter side acknowledges the transfer */
+    do {
+        err = mhu_v3_x_doorbell_read(dev, ch, &read_val);
+        if (err != MHU_V_3_X_ERR_NONE) {
+            goto fail;
+        }
+    } while ((read_val & value) == value);
+
+fail:
+    return err;
+}
+
+static enum mhu_v3_x_error_t wait_for_signal_and_clear(
+        struct mhu_v3_x_dev_t *dev, uint32_t ch, uint32_t value)
+{
+    enum mhu_v3_x_error_t err = MHU_V_3_X_ERR_NONE;
+    uint32_t read_val;
+
+    /* Wait on status register for transmitter side to send data */
+    do {
+        err = mhu_v3_x_doorbell_read(dev, ch, &read_val);
+        if (err != MHU_V_3_X_ERR_NONE) {
+            goto fail;
+        }
+    } while ((read_val & value) != value);
+
+    /* Acknowledge the transfer and clear the doorbell register */
+    err = mhu_v3_x_doorbell_clear(dev, ch, value);
+    if (err != MHU_V_3_X_ERR_NONE) {
+        goto fail;
+    }
+
+fail:
+    return err;
+}
+
+enum mhu_error_t mhu_send_data(void *mhu_sender_dev, void *mhu_outband_base,
+        uint8_t *send_buffer, size_t size)
+{
+    struct mhu_v3_x_dev_t *dev = mhu_sender_dev;
+    uint32_t msg_len = (uint32_t)size;
+    enum mhu_v3_x_error_t err;
+
+    if (size == 0) {
+        return error_mapping_to_mhu_error_t(MHU_V_3_X_ERR_NONE);
+    }
+
+    err = validate_buffer_params((uintptr_t)send_buffer, size);
+    if (err != MHU_V_3_X_ERR_NONE) {
+        goto fail;
+    }
+
+    /*
+     * First 4 bytes represents the header which currently stores the size
+     * of msg. Rest of buffer holds the actual message.
+     */
+    /* (void)memcpy((void *)MHU3_OUTBAND_BUF_BASE, (void *)&msg_len,
+            MHU3_OUTBAND_BUF_HEADER_SIZE); */
+
+    /* Copy the message */
+    /* (void)memcpy((void *)(MHU3_OUTBAND_BUF_BASE + MHU3_OUTBAND_BUF_HEADER_SIZE),
+            (void *)send_buffer, msg_len); */
+
+    /*
+     * Use only one channel 0 to ring the doorbell on receiver for
+     * attestation report and measurements, while the actual data is
+     * shared through a common memory region in Secure/Root PAS. After
+     * receiver is done with this transaction, it needs to clear the flags
+     * bits on the same channel or else transmitter will continue waiting
+     * on the status bit indefinitely.
+     */
+    err = signal_and_wait_for_clear(dev, 0, MHU3_PBX_DBCH_FLAG_AP_COMMS);
+    if (err != MHU_V_3_X_ERR_NONE)
+        goto fail;
+
+    return MHU_V_3_X_ERR_NONE;
+
+fail:
+
+    return error_mapping_to_mhu_error_t(err);
+}
+
+enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
+        uint8_t *receive_buffer, size_t *size)
+{
+    struct mhu_v3_x_dev_t *dev = mhu_receiver_dev;
+    uint32_t msg_len;
+    enum mhu_v3_x_error_t err;
+
+    err = validate_buffer_params((uintptr_t)receive_buffer, *size);
+    if (err != MHU_V_3_X_ERR_NONE) {
+        goto fail;
+    }
+
+    /*
+     * Use only one channel 0 to receive reply from transmitter for
+     * attestation report and measurements, while the actual data is shared
+     * through a common memory region in Secure/Root PAS.
+     */
+    err = wait_for_signal_and_clear(dev, 0, MHU3_PBX_DBCH_FLAG_AP_COMMS);
+    if (err != MHU_V_3_X_ERR_NONE) {
+        goto fail;
+    }
+
+    /*
+     * First 4 bytes represents the header which currently stores the size
+     * of msg. Rest of buffer holds the actual message.
+     */
+    /* (void)memcpy((void *)&msg_len, (void *)MHU3_OUTBAND_BUF_BASE,
+            MHU3_OUTBAND_BUF_HEADER_SIZE);
+    if (*size < msg_len) {
+        return MHU_V_3_X_ERR_UNSUPPORTED;
+    } */
+
+    *size = msg_len;
+
+    /* Copy the message */
+    /* (void)memcpy((void *)receive_buffer, (void *)(MHU3_OUTBAND_BUF_BASE +
+                MHU3_OUTBAND_BUF_HEADER_SIZE), msg_len); */
+
+    return MHU_V_3_X_ERR_NONE;
+
+fail:
+
+    return error_mapping_to_mhu_error_t(err);
+}
-- 
2.25.1

